File: epi_judge_cpp/buy_and_sell_stock.cc
#include <vector>
#include <algorithm>

#include "test_framework/generic_test.h"
using namespace std;
double BuyAndSellStockOnce(const vector<double> &prices)
{
    int n = prices.size();

    if (n == 0) {
        return 0;
    }

    double lowest_so_far = prices[0];
    double max_profit = 0;

    for (int i = 1; i < n; i++) {
        max_profit = max(max_profit, prices[i] - lowest_so_far);
        lowest_so_far = min(lowest_so_far, prices[i]);
    }
    
    return max_profit;
}

int main(int argc, char *argv[])
{
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"prices"};
    return GenericTestMain(args, "buy_and_sell_stock.cc",
                           "buy_and_sell_stock.tsv", &BuyAndSellStockOnce,
                           DefaultComparator{}, param_names);
}

File: epi_judge_python/buy_and_sell_stock.py
from typing import List

from test_framework import generic_test


def buy_and_sell_stock_once(prices: List[float]) -> float:
    n = len(prices)
    max_profit = 0
    lowest_price_before_i = prices[0]

    for i in range(1, n):
        max_profit_selling_i = prices[i] - lowest_price_before_i
        max_profit = max(max_profit, max_profit_selling_i)
        lowest_price_before_i = min(lowest_price_before_i, prices[i])

    return max_profit


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('buy_and_sell_stock.py',
                                       'buy_and_sell_stock.tsv',
                                       buy_and_sell_stock_once))
